web/profiles/joinup/src/Form/SolutionsByLicenceForm.php

```diff
@@ -0,0 +1,240 @@
+
+use Drupal\Core\Form\FormBase;
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\Core\Url;
+use Drupal\rdf_entity\Database\Driver\sparql\ConnectionInterface;
+use Drupal\rdf_entity\Entity\Query\Sparql\SparqlArg;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * A simple page that presents a TCA form for the collection creation.
+ */
+class SolutionsByLicenceForm extends FormBase {
+
+  /**
+   * The items per page.
+   */
+  const ITEMS_PER_PAGE = 50;
+
+  /**
+   * The SPARQL connection.
+   *
+   * @var \Drupal\rdf_entity\Database\Driver\sparql\ConnectionInterface
+   */
+  protected $connection;
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getFormId() {
+    return 'solutions_by_licence_form';
+  }
+
+  /**
+   * SolutionsByLicenceController constructor.
+   *
+   * @param \Drupal\rdf_entity\Database\Driver\sparql\ConnectionInterface $connection
+   *   The SPARQL connection.
+   */
+  public function __construct(ConnectionInterface $connection) {
+    $this->connection = $connection;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container) {
+    return new static(
+      $container->get('sparql_endpoint')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildForm(array $form, FormStateInterface $form_state) {
+    if (empty($form_state->get('licence_options'))) {
+      $form_state->set('licence_options', $this->getLicenceOptions());
+    }
+
+    $form['licence_option'] = [
+      '#type' => 'select',
+      '#title' => $this->t('Licence'),
+      '#options' => $form_state->get('licence_options'),
+      '#default_value' => $form_state->getValue('licence_option'),
+      '#empty_option' => ' - All - ',
+    ];
+    $form['filter'] = [
+      '#type' => 'submit',
+      '#value' => $this->t('Filter'),
+    ];
+
+    $licence_id = $form_state->getValue('licence_option');
+    $results = $this->connection->query($this->getCountQuery($licence_id));
+    $total = reset($results)->total->getValue();
+    $page = pager_default_initialize($total, self::ITEMS_PER_PAGE);
+    $offset = $page * self::ITEMS_PER_PAGE;
+    $items = $this->connection->query($this->getQuery($offset, $licence_id));
+
+    $rows = [];
+    foreach ($items as $item) {
+      $licence_url = Url::fromRoute('entity.rdf_entity.canonical', [
+        'rdf_entity' => $item->licence,
+      ]);
+      $solution_url = Url::fromRoute('entity.rdf_entity.canonical', [
+        'rdf_entity' => $item->solution,
+      ]);
+
+      $rows[] = [
+        $this->getLinkGenerator()->generate($item->licence_label, $licence_url),
+        $this->getLinkGenerator()->generate($item->solution_label, $solution_url),
+      ];
+    }
+
+    $form['table'] = [
+      '#theme' => 'table',
+      '#header' => ['Licence', 'Solution'],
+      '#rows' => $rows,
+      '#empty' => $this->t('No solutions available.'),
+    ];
+    $form['pager'] = [
+      '#type' => 'pager',
```

Comment:
This pager is not being tested, but that's OK for me. This is a moderator-only functionality which is not usable by the public. I think it is not necessary to spend the testing time for creating a matrix of 50+ licenses and solutions for this feature which will probably be very little used.

Reply:
Actually that was the only reason I did not test it. I did test manually the extend of the pager and if all pages were ok, but to write a test with 50+ cases just for the pager seemed too much for me :(